### DISCUSSION TITLE
Noise implementation

### DISCUSSION THREAD

**Discussion by yudhistiraindyka** [thejoabo/virtualfisher-bot#12] [Ideas]
Hello @thejoabo, I tried looking at the code and I can't really understand how you implement noise for cooldown in the 1.0 and later updates, can you explain roughly how it works? Also I found this a while ago on Virtual Fisher discord server

> The timing of most commands, both in individual users and in the server at large, follows a rough bell curve. A long time ago, I calculated that the average length between commands is approximately 2.15 seconds, and the deviation in command times can be described by the function f(x) = 0.5e^-(52x^2) where the x-axis represents variation in seconds and the y axis represents frequency of timing. Integration of this bell curve from negative to positive infinity yields a value of ~0.12289; this value represents the weighted variation in times between commands. For determining whether a player is a botter or not, the bot takes the time between their commands and makes a bell curve with this data. This bell curve is integrated with negative and positive infinite bounds, and if the value lies outside ±0.3 from the sample bell curve integral, that player is flagged as suspicious. If the player is flagged, a pigeonhole sorting algorithm is applied to their times, and a Taylor series approximation centered at x=0 is created using the player's time between commands. If a Lagrange error bound approximation yields a statistically significant truncation error, calculated using a Chi-Square value between their real times and this polynomial approximation (with 3 degrees of freedom and critical value p=0.05,) then they are automatically banned.

I don't really understand what those means but it might help noise implementation for future updates

**Comment by thejoabo** [comment]
My noise implementation was made on the first version of this autofisher, and to be honest I didn't even worried to much about it. I just figure that if the command is sent at a constant rate (+ deviation from requests) it would be pretty easy to spot bots. 
So I implemented a function where it takes the user's cooldown and adds a random value from the ranges 0 to 1.500 + standard small delay between requests. For me that was good enough since it would simulate a variation and mimic (poorly, to say the least) a legit player.

As you can see, I'm not really familiar to this community and don't have a lot knowledge about how everything works. But thanks for sharing this message I didn't know that a simple Virtual Fisher would present this fancy math and shit. 

Well, I can try to implement a method to scramble a bit with this type of analysis.

Btw, the new version (1.1.1) solved the prestige regex problem ?

**Comment by yudhistiraindyka** [comment]
Yeah the regex problem is solved, thank you very much!

**Comment by thejoabo** [comment]
Hey @yudhistiraindyka,
I was just playing around with this issue and I think I found a proper solution to bypass this type of analysis. Keep in mind that I'm not that good when it comes to math, so I'll probably just reach out to someone that knows about statistics and ask for help.

### My assumptions:
- since they supposedly analyzed a large amount of data, I assumed that the standard deviation of any user tends to ~0.12289, so this value should be the _sigma_;
- the average length of this data is 2.15 seconds, so I'm assuming that is the _mean_ (usual average) so following the assumption above, if I use the USER_COOLDOWN as the _mean_ it should tend to that _sigma_ over time;
- the Gaussian distribution is generated per-user and not compared with some other data ("_the bot takes the time between their commands and makes a bell curve with this data._")

### My solution:
First of all, I should generate a new Gaussian distribution using the USER_COOLDOWN as _mean_ and that constant value (0.12289) as _sigma_, with a size of 100 values (and I'm not sure if its really necessary the 100 values or just 1 is enough, but maybe is worth to check the overall mean variation of this new list of values). Then I would pick a random value of this distribution and use as next command cooldown. This way, when analyzed, and their bot generates a bell curve using the user's data they would get a legitimate curve and it shouldn't raise any suspicion.

### Proof:
So let's suppose that after that pigeonhole sorting algorithm the selected the values from the "%f" command. This is the expected curve using the _mean = 3.5_ (user cd) and _sigma = 0.12289_ (deviation constant), **this sample contains one million values**:

![curve](https://user-images.githubusercontent.com/49594896/180630342-cd78cfbd-619e-4f74-8085-2ded116b30f7.png)
_If all the assumptions are correct, this should be a legitimate curve since it follows all the proprieties of this kind of distribution ._

To generate a single value contained in this set is pretty simple:
`cdlist = np.random.normal(loc=user_cd, scale=sigma, size=100)`
`cooldown = cdlist[randint(0, len(cdlist))]`
`Output (for this one): cooldown = 3.5878046897169154`

Anyways, I should implement this method in the next update, and try to improve it. **What do you think about this approach ?** For me, at least, makes a lot of sense.

**Comment by yudhistiraindyka** [comment]
Hello @thejoabo, I think your approach really does makes sense and helps me a lot to understand the analysis, even though I'm not really good at math myself. Since it might be apparent now that I'm also not that good at programming, my contributions would most probably be some form of ideas for features that are yet to be implemented, bug testing, or keeping up with the community to stay updated. Thank you so much.

**Reply by thejoabo** [reply]
No problem at all. Your help with ideas is very much appreciated, please feel free to suggest more features (I'm not a creative person myself lol).
